File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        int n = queue1.length;

        long sum1 = 0;
        long sum2 = 0;

        Deque<Integer> que1 = new ArrayDeque<>();
        Deque<Integer> que2 = new ArrayDeque<>();

        for (int v : queue1) {
            que1.offer(v);
            sum1 += v;
        }

        for (int v : queue2) {
            que2.offer(v);
            sum2 += v;
        }

        if ((sum1 + sum2) % 2 == 1) {
            return -1;
        }

        for (int i = 0; i < 3 * n; i++) {
            if (sum1 == sum2) {
                return i;
            } else if (sum1 > sum2) {
                int value = que1.poll();
                que2.offer(value);
                sum1 -= value;
                sum2 += value;
            } else {
                int value = que2.poll();
                que1.offer(value);
                sum2 -= value;
                sum1 += value;
            }
        }

        return -1;
    }
}

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.swift
import Foundation

func solution(_ queue1: [Int], _ queue2: [Int]) -> Int {
    var que1: Queue<Int> = Queue()
    var que2: Queue<Int> = Queue()
    
    let n = queue1.count
    var q1sum = 0
    var q2sum = 0
    
    
    for value in queue1 {
        que1.enqueue(value)
        q1sum += value
    }
    
    for value in queue2 {
        que2.enqueue(value)
        q2sum += value
    }
    
    for i in 0..<n*3 {
        if q1sum == q2sum {
            return i
        }
        else if q1sum > q2sum {
            guard let value = que1.dequeue() else {return -1}
            que2.enqueue(value)
            
            q1sum -= value
            q2sum += value
        }
        else {
            guard let value = que2.dequeue() else {return -1}
            que1.enqueue(value)
            
            q1sum += value
            q2sum -= value
        }
    }
    
    return -1;
}

struct Queue<T> {
    private var inbox : [T] = []
    private var outbox : [T] = []
    
    var isEmpty : Bool {
        return inbox.isEmpty && outbox.isEmpty
    }
    
    mutating func enqueue(_ element: T) {
        inbox.append(element)
    }
    
    mutating func dequeue() -> T? {
        if(outbox.isEmpty) {
            outbox = inbox.reversed()
            inbox.removeAll()
        }
        
        return outbox.popLast()
    }
}